scio-core/src/main/scala/com/spotify/scio/coders/CoderMaterializer.scala

```diff
@@ -39,62 +41,65 @@ object CoderMaterializer {
     o: PipelineOptions = PipelineOptionsFactory.create()
   ): BCoder[T] = beam(o, coder)
 
-  @inline private def nullCoder[T](o: CoderOptions, c: BCoder[T]) =
-    if (o.nullableCoders) NullableCoder.of(c)
-    else c
-
   final def beam[T](
     o: PipelineOptions,
     coder: Coder[T]
-  ): BCoder[T] = beamImpl(CoderOptions(o), coder)
+  ): BCoder[T] = beamImpl(CoderOptions(o), coder, topLevel = true)
+
+  private def isNullableCoder(o: CoderOptions, c: Coder[_]): Boolean = c match {
+    case _: RawBeam[_]      => false // raw cannot be made nullable
+    case _: KVCoder[_, _]   => false // KV cannot be made nullable
+    case _: Transform[_, _] => false // nullability should be deferred to transformed coders
+    case _: Ref[_]          => false // nullability should be deferred to underlying coder
+    case _                  => o.nullableCoders
+  }
+
+  private def isWrappableCoder(topLevel: Boolean, c: Coder[_]): Boolean = c match {
+    case _: RawBeam[_]    => false // raw should not be wrapped
+    case _: KVCoder[_, _] => false // KV should not be wrapped, but independent k,v can
+    case _                => topLevel
```

Comment:
Same here. Please avoid refection match.

Reply:
You can do this on sealed trait without issues. No need to use `unapply`. Eg. if I remove the default case, the compiler tells
```sh
 match may not be exhaustive.
[warn] It would fail on the following inputs: Beam(_), Disjunction(_, _, _, _), Fallback(_), Record(_, _, _, _), Ref(), Transform(_, _)
```

Reply:
It'll still compile to runtime reflection which I'd try to avoid.

Reply:
I'm not sure I follow. The scala compiler will translate those in both cases with an `isInstanceOf` check regardless if pattern match is used with the type or the `unapply`.

Reply:
Looks like you're right :) It does not matter then.